data_model/src/query/predicate/mod.rs

```diff
@@ -0,0 +1,370 @@
+                account::{AccountIdPredicateBox, AccountPredicateBox},
+                domain::DomainIdPredicateBox,
+                PublicKeyPredicateBox,
+            },
+            projectors::BaseProjector,
+            prototypes::account::AccountPrototype,
+            CompoundPredicate,
+        },
+    };
+
+    /// Smoke test that creates a simple predicate using a prototype
+    #[test]
+    fn test_prototype_api() {
+        let alice_account_id: AccountId =
+            "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland"
+                .parse()
+                .unwrap();
+
+        // construct a prototype (done by the machinery)
+        let account = AccountPrototype::<BaseProjector<AccountPredicateBox>>::default();
+        // take a look at the type name (it should be `AccountIdPrototype<AccountIdProjector<BaseProjector<AccountPredicateBox>>>`)
+        #[allow(unused)]
+        let account_id_prototype = account.id;
+        // make a predicate from it (done by the user)
+        let predicate = account.id.eq(alice_account_id.clone());
+        // normalize it (done by the machinery)
+        let predicate_norm = predicate.normalize();
+
+        // check that the predicate is correct
+        assert_eq!(
+            predicate_norm,
+            CompoundPredicate::Atom(AccountPredicateBox::Id(AccountIdPredicateBox::Equals(
+                alice_account_id
+            )))
+        );
+    }
+
+    /// Same as [`test_prototype_api`], but uses the `AccountPredicateBox::build()` method
+    #[test]
+    fn test_builder_api() {
+        let alice_account_id: AccountId =
+            "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland"
+                .parse()
+                .unwrap();
+
+        let predicate_norm =
+            AccountPredicateBox::build(|account| account.id.eq(alice_account_id.clone()));
+
+        // check that the predicate is correct
+        assert_eq!(
+            predicate_norm,
+            CompoundPredicate::Atom(AccountPredicateBox::Id(AccountIdPredicateBox::Equals(
+                alice_account_id
+            )))
+        );
+    }
+
+    /// Create a denormalized predicate (logical combinator inside a projection), check that it normalizes correctly
+    #[test]
+    fn test_prototype_normalization() {
+        let alice_signatory: PublicKey =
+            "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03"
+                .parse()
+                .unwrap();
+        let alice_domain_id: DomainId = "wonderland".parse().unwrap();
+
+        let account_predicate_denorm = AccountPredicateBox::build_fragment(|account| {
+            let account_id_predicate = AccountIdPredicateBox::build_fragment(|account_id| {
+                account_id
+                    .signatory
+                    .eq(alice_signatory.clone())
+                    .and(account_id.domain_id.eq(alice_domain_id.clone()))
```

Comment:
You mentioned in the PR description the use of `impl Into<T>` to reduce clutter. I guess it will also apply to these methods (if implemented in future)

```suggestion
                    .eq(&alice_signatory)
                    .and(account_id.domain_id.eq(&alice_domain_id))
```

Reply:
It is only a future work, I haven't actually done this for non-string types =(